### PR TITLE
daml-sdk-head: add optional sha information

### PR DESCRIPTION
**Pull request by garyverhaegen-da**
This PR add an option, `--sha`, to `daml-sdk-head` so that it produces a more accurate version number including the current git sha.

The main consequence is that calling `daml-sdk-head --sha` take significantly longer than calling `daml-sdk-head`, because it needs to recompile everything that depends on the version number.

> ## Wait, but why??

I started this work in support of an internal project that needed to test against unreleased, and possibly unmerged, daml versions. However, after further discussion I believe there is a better option for their use-case. I've decided to still open this PR because the work was done and there is no downside to it. It may still be useful if one wanted to be able to maintain more than one non-released local version of daml.

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

**Comment by garyverhaegen-da** [comment]
I should note that I will not be offended if the overall consensus ends up being that this should not be merged.

File: bazel_tools/build_environment.bzl [hunk 1]
```diff
@@ -1,6 +1,15 @@
 # Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
 # SPDX-License-Identifier: Apache-2.0
 
+def is_figure(c):
+    return c == "0" or c == "1" or c == "2" or c == "3" or c == "4" or c == "5" or c == "6" or c == "7" or c == "8" or c == "9"
```
Comment (SamirTalwar-DA) [inline bazel_tools/build_environment.bzl:5]: I suggest using `in`.

```suggestion
valid_figures = "0123456789"
def is_figure(c):
    return c in valid_figures
```

File: bazel_tools/build_environment.bzl [hunk 2]
```diff
@@ -12,11 +21,17 @@ def _impl(ctx):
     # Generates a simple Bazel file that just sets a bunch of Bazel variables,
     # so they can be used in our main Bazel BUILD files.
     semver = ctx.os.environ.get("DAML_SDK_RELEASE_VERSION", default = "0.0.0")
-    if semver.find("-snapshot.") > 0:
-        ghc = semver[:-9].replace("-snapshot.", ".")
+    if semver.find("-") > 0:
+        ghc_list = []
+        for segment in semver.replace("-", ".").split("."):
+            if is_number(segment):
```
Comment (cocreature) [inline bazel_tools/build_environment.bzl:27]:
```suggestion
            if segment.isdigit():
```
and remove all the stuff above.

File: bazel_tools/build_environment.bzl [hunk 3]
```diff
@@ -12,11 +21,17 @@ def _impl(ctx):
...
+                ghc_list.append(segment)
+        ghc = ghc_list[0]
+        for elem in ghc_list[1:]:
+            ghc = ghc + "." + elem
```
Comment (cocreature) [inline bazel_tools/build_environment.bzl:31]:
```suggestion
        ghc = ".".join(ghc_list)
```

File: bazel_tools/build_environment.bzl (hunk 1 above)
Comment (SamirTalwar-DA) [reply bazel_tools/build_environment.bzl:5]: And perhaps "digit" instead of "figure".

File: bazel_tools/build_environment.bzl [hunk 4]
```diff
@@ -1,6 +1,15 @@
...
 # SPDX-License-Identifier: Apache-2.0
 
+def is_figure(c):
+    return c == "0" or c == "1" or c == "2" or c == "3" or c == "4" or c == "5" or c == "6" or c == "7" or c == "8" or c == "9"
+
+def is_number(s):
+    for c in s.elems():
+        if not is_figure(c):
+            return False
+    return True
```
Comment (SamirTalwar-DA) [inline bazel_tools/build_environment.bzl:11]: You can use `all` here.

```suggestion
def is_number(s):
    return all([is_figure(c) for c in s.elems()])
```

File: dev-env/bin/daml-sdk-head [hunk 5]
```diff
@@ -58,7 +64,21 @@ if [[ -d "$DAML_HOME" && "$NUKE" -ne 0 ]]; then
   rm -rf "$DAML_HOME"
 fi
 
-readonly DAML_HEAD_SDK="$DAML_HOME/sdk/0.0.0"
+sha_version() {
+    local dirty date nr sha8
+    dirty=$([[ -z $(git status -s) ]] && echo clean || echo dirty)
```
Comment (cocreature) [inline dev-env/bin/daml-sdk-head:69]: The way I’ve usually seen these kinds of version numbers build up is to include `dirty` if it’s dirty and nothing if it’s not dirty. Any reason why we want `clean` in there explicitly?

**Review by cocreature — APPROVED** [review]
LGTM with a bit of starlark bikeshedding, thank you!

File: bazel_tools/build_environment.bzl (hunk 3 above)
Comment (SamirTalwar-DA) [inline bazel_tools/build_environment.bzl:31]:
```suggestion
        ghc_list = [segment for segment in semver.replace("-", ".").split(".") if is_number(segment)]
        ghc = ".".join(ghc_list)
```

**Review by SamirTalwar-DA — COMMENTED** [review]
Got some Bazel comments. Gonna review the rest now.

File: bazel_tools/build_environment.bzl (hunk 2 above)
Comment (garyverhaegen-da) [reply bazel_tools/build_environment.bzl:27]: Oh, missed that.

File: bazel_tools/build_environment.bzl (hunk 3 above)
Comment (garyverhaegen-da) [reply bazel_tools/build_environment.bzl:31]: I was looking for a method on list. 🤦

File: dev-env/bin/daml-sdk-head (hunk 5 above)
Comment (SamirTalwar-DA) [reply dev-env/bin/daml-sdk-head:69]: If you want a truly meaningful version number, you may want to hash the tree if it's dirty. spider-on-x does this.

1. [hash-repository.sh](https://github.com/DACH-NY/spider-on-x/blob/master/helpers/hash-repository.sh)
2. [hash-tree.sh](https://github.com/DACH-NY/spider-on-x/blob/master/helpers/hash-tree.sh)

Comment (garyverhaegen-da) [reply dev-env/bin/daml-sdk-head:69]: Not really, no. I'll update.

File: bazel_tools/build_environment.bzl (hunk 4 above)
Comment (garyverhaegen-da) [reply bazel_tools/build_environment.bzl:11]: I was looking for `map` and was really sad not to find it. I did not even think of comprehension. I guess my Python is getting very rusty at this point.

Comment (SamirTalwar-DA) [reply bazel_tools/build_environment.bzl:11]: Ah well. @cocreature's solution is much better. :grinning:

File: bazel_tools/build_environment.bzl (hunk 3 above)
Comment (SamirTalwar-DA) [reply bazel_tools/build_environment.bzl:31]: IMO Python is the only language that does this "correctly". :stuck_out_tongue:

File: bazel_tools/build_environment.bzl (hunk 4 above)
Comment (garyverhaegen-da) [reply bazel_tools/build_environment.bzl:11]: Actually I think I can one-up that.

**Review by SamirTalwar-DA — APPROVED** [review]

**Pull request merged**